### ISSUE TITLE
Binary props vs enum

### ISSUE THREAD

**Issue by siddharthkp** [auth0/cosmos#88]
What makes a more intuitive API?

```jsx
<Button primary large>Button</Button>

<Button transparent small>Button</Button>
<Button destructive small>Button</Button>
```

	vs.

```jsx
<Button type="primary" size="large">Button<Button>

<Button type="transparent" size="small">Button</Button>
<Button type="destructive" size="small">Button</Button>
```

The first approach has a simpler API, you write as you think, _I need a large primary button_, the conflicting props are not obvious, you need to rely on `prop-type` validation warnings to catch mistakes

![screen shot 2017-12-22 at 6 30 45 pm](https://user-images.githubusercontent.com/1863771/34312920-4c4677e8-e746-11e7-9fa1-5aea1e40da5d.png)

The second approach does not have conflict problems at the cost of a longer/more explicit API.

**Comment by luisrudge** [comment]
I vote for the first one 😬

**Comment by nicogarcia** [comment]
At first glance I like the first one but does it allow to dynamically set the value as does the second? E.g. `<Button type={buttonType} ...`

**Comment by luisrudge** [comment]
it does.

```js
<Button
    transparent={shouldBeTransparent}
    small>Button</Button>
```

**Comment by nicogarcia** [comment]
Yeah, but that should be `<Button transparent={shuldBeTransparent} primary={shouldBePrimary} secondary={shouldBeSecondary}...` I think, for some use cases. Maybe both approaches can be accepted? Or would it be confusing?

**Comment by astanciu** [comment]
I too vote for the second one. You have one property (type), with many options (primary, transparent, etc..)

The first format would allow me to do bad things like:
```JSX
<Button primary transparent>Button</Button>
```

What's the type there?

**Comment by luisrudge** [comment]
> Yeah, but that should be <Button transparent={shuldBeTransparent} primary={shouldBePrimary} secondary={shouldBeSecondary}... I think, for some use cases. Maybe both approaches can be accepted? Or would it be confusing?

Well, this is the code for your example:

```js
<Button type={shuldBeTransparent ? 'transparent' : shouldBePrimary ? 'primary' : shouldBeSecondary ? 'secondary' : 'default'} />
```

I rather have the first one 😊

**Comment by alejofernandez** [comment]
I vote for option 1, I wasn't a fan of it when I saw it first, but I'm convinced now that it is more readable.

**Comment by ramasilveyra** [comment]
I vote for `<Button primary large>` approach.

**Comment by siddharthkp** [comment]
@alejofernandez @ramasilveyra @luisrudge @nicogarcia thanks!

@astanciu That's exactly the problem with it, it's easy to miss/be confused.

@alejofernandez what made you change your mind?

@nicogarcia Is that a common use case? I've usually come across buttons which are static, they were not based on data. I've seen forms which have dynamically generated fields but actions/buttons were still constant

**Comment by nicogarcia** [comment]
@siddharthkp That was my question in the first comment, I wanted to be sure to avoid disabling some use cases, but as this component may not be for dynamic use maybe it's better to bet for readability, ease of use and autocompletion, so option 1 would be better.

**Comment by siddharthkp** [comment]
Ah, okay sorry. I see your point though.

This could be an annoying problem with `Input` especially because it already has the HTML `type` property.

```jsx
<Input type="password" />
<Input type="number" />
<Input type="code" /> // custom non-semantic code?
```

vs

```jsx
<Input password /> // wrappers around semantic HTML?
<Input number />
<Input code />
```

**Comment by luisrudge** [comment]
Why not expose different inputs? PasswordInput, NumberInput, CodeInput etc?

**Comment by siddharthkp** [comment]
> Why not expose different inputs? PasswordInput, NumberInput, CodeInput etc?

That could work, seems very similar to `<Input password/>` approach?

**Comment by luisrudge** [comment]
Yeah.. it's similar but prevents the Input component of being too big (imagine having all types inside just a single component: text, number, password, code, color, range etc)

**Comment by siddharthkp** [comment]
First impressions:

`<Button primary large>Button</Button>`: 4 (@luisrudge @alejofernandez @ramasilveyra @siddharthkp)

`<Button type="primary" size="large">Button<Button>`: 4 (@nicogarcia @astanciu @landitus @nkohari )

This seems to be the thing that hits immediately:

> The first format would allow me to do bad things like:

```
<Button primary transparent>Button</Button>
```

There is a warning, but I'm not sure if that's going to be good enough.

Open to suggestions about improving/changing the API

**Comment by nkohari** [comment]
I ran into a situation working with `DialogActions` that might add more data to this decision. I have a `DialogAction` like:

```
class DialogAction {
  constructor(label, method, appearance = "default") {
    this.label = label
    this.method = method
    this.appearance = appearance
  }
}
```

Then in `Dialog`, this action gets translated into a `Button`. If `Button` accepted an `appearance` prop (for example), I could pass the value of `DialogAction.appearance` through to the `Button` as a prop. Because it requires booleans, I have to either create a switch statement or do a "creative" 😅  hack using object interpolation, like this:

```
const createButtonForAction = action => {
  const buttonProps = {
    onClick: action.method,
    [action.apperance]: true
  }
  return <Button {...buttonProps} />
}
```

Maybe there's another way to do this, but I thought this was an interesting tension point in the current `Button` API.

**Comment by beneliflo** [comment]
Vote for second one `<Button type="primary" size="large">Button<Button>`
But instead of using `type =""` we can use `appearance = ""` maybe. `type` in HTML already has a unique meaning.

**Comment by nkohari** [comment]
FWIW, I've also seen this called `intent` in a couple of component libraries. Not sure if that's better or worse.

**Comment by landitus** [comment]
Other options are `buttonStyle` which sounds more presentational and maybe it gives out the wrong idea that you could pass CSS inside that prop. 

Like the idea of avoiding common HTML in the props. I like `intent`, which is good as it gives out the meaning in a good and short word. Intent primary sounds good and it's easy to write.

**Comment by alejofernandez** [comment]
`Just a thought, and completely off-topic to the issue we are discussing here`

I would like to understand a little better the main motivation of having a generic DialogAction class, I'm not sure if it is a good idea. It is one of those cases in which the inheritance does not help a lot.
Specially because propagating the value of your props to inner components breaks the principle of encapsulation.
I would rather have different types of dialog actions, which serves more to the purpose of standardizing, ie: `OkDialogAction`, `CancelDialogAction`, `YesDialogAction`, `NoDialogAction`, and each one of them has the responsibility of displaying a label and a button (it can be primary, transparent, whatever), and receiving an event handler. The pro of this approach would be that you'll always show the Cancel button in the same way.

**Comment by nkohari** [comment]
If we wanted a closed API rather than an open one, I'd just implement it using an enum rather than subclasses. Because there isn't any different *behavior* between the different subclasses, I think using inheritance is overkill in this case. An example of a closed API would be:

```jsx
class MyDialog extends React.Component {
  handleOkClicked = () => { ... }
  handleCancelClicked = () => { ... }
  render() {
    const actions=[{
      Dialog.action.OK(this.handleOkClicked),
      Dialog.action.Cancel(this.handleCancelClicked)
    ];
    return (
      <Dialog actions={actions}>
        ...
      </Dialog>
    );
  }
}
```

The `Dialog.action.OK()` (etc.) methods would just create an instance of `DialogAction` rather than calling the constructors directly, and the `DialogAction` class wouldn't be exported outside of the `Dialog` package.

**Comment by alejofernandez** [comment]
How about this? Doesn't it look more readable?
```
class MyDialog extends React.Component {
  handleOkClicked = () => { ... }
  handleCancelClicked = () => { ... }
  render() {
    return (
      <Dialog actions={actions}>
        ...
         <OKDialogAction action={() => this.handleOkClicked} />
         <CancelDialogAction action={() => this.handleCancelClicked} />
      </Dialog>
    );
  }
}
```

**Comment by nkohari** [comment]
Hmm... well, that's not what the API will look like, since we can't just mix the actions in with the rest of the content of the `Dialog`. It'd be more like:

```jsx
class MyDialog extends React.Component {
  handleOkClicked = () => { ... }
  handleCancelClicked = () => { ... }
  render() {
    return (
      <Dialog actions={actions}>
        ...
        <Dialog.Actions>
          <OKDialogAction action={() => this.handleOkClicked} />
          <CancelDialogAction action={() => this.handleCancelClicked} />
        </Dialog.Actions>
      </Dialog>
    );
  }
}
```

Which means there's another component (`Dialog.Actions`) that consumers have to understand how to use (and that we have to document!) which would otherwise have been internal.

Also, actions are now used in at least 3 places (Form, Dialog, and Form.TextInput / ActionInput), so I think it might be valuable to see if we can separate out the concept of "action" from any particular representation. For example, we might have:

```jsx
const actions = [
  Action.primary(this.handleSave, { label: "Save Changes" }),
  Action.secondary(this.handleClear, { label: "Clear" }),
  Action.destructive(this.handleDelete, { label: "Delete Client", icon: "delete" })
]
return (
  <Form actions={actions}>
    ...
  </Form>
)
```

This would result in 3 buttons being displayed at the end of the form. But, then we could also use the same API to render a `Form.TextInput`:

```jsx
const actions = [
  Action.primary(this.handleCopy, { label: "Copy to clipboard", icon: "copy" })
]
return (
  <Form>
    <Form.TextInput actions={actions} ... />
    ...
  </Form>
)
```

...and in this case the input would be rendered with a single icon button with the tooltip "Copy to clipboard".

Long story short, I think that users shouldn't directly interact with actions as components, but instead with the concept of actions as part of the parent component (Form, etc.) itself.

**Comment by alejofernandez** [comment]
Agree, sounds like a good abstraction.
